src/Irc.hs

```diff
@@ -39,25 +39,32 @@ import           Network.IRC.CTCP             (CTCPByteString, asCTCP)
 import           Prelude                      hiding (and)
 import           System.Console.Concurrent    (outputConcurrent)
 import           System.IO.Error              (ioeGetErrorString)
+import           Control.Exception.Lifted
 
 import           Network.SimpleIRC
 
-connectTo :: IrcParams -> Bool -> IO () -> IO () -> IrcIO Connection
-connectTo params withDebug onConnected onJoined = do
+connectTo :: Options -> IrcParams -> Bool -> IO () -> IO () -> IrcIO Connection
+connectTo opt@Options{..} params withDebug onConnected onJoined = do
     (conf, bcs) <- lift $ config params
     con <- connect' conf withDebug
     lift $ mapM_ ((\f -> f con) . onConnect) (hooks params)
     lift onConnected
-    joined <- lift $ waitForAll bcs
+    joined <- lift $ catch (waitForAll bcs) (handler con)
```

Comment:
On Sat, 16 Apr 2016 10:25:46 -0700
Jan Gerlinger <notifications@github.com> wrote:

> >      (conf, bcs) <- lift $ config params
> >      con <- connect' conf withDebug
> >      lift $ mapM_ ((\f -> f con) . onConnect) (hooks params)
> >      lift onConnected
> > -    joined <- lift $ waitForAll bcs
> > +    joined <- lift $ catch (waitForAll bcs) (handler con)
> 
> Wouldn't an exception here be handled and rethrown by the `bracket`
> that ultimately calls this function which then releases the
> connection?

bracket only guarantees that once a resource has been
successfully acquired, any exception thrown in the 'apply' function
will be caught and handled. 

The code above however runs in the 'acquire' function, hence any
exception thrown will be propagated upwards immediately and 'release'
is never invoked.

The reason i used 'catch' for 'waitForAll' was that while waiting for
znc to join channels after connecting, i pressed Ctrl-C causing an
async exception that wouldn't be caught by bracket. This resulted in
znc never receiving a 'disconnect' message because bracket's 'release'
was never invoked.